ENS usernames depend on the network

igor-sirotin opened the pull request:
Fixes #3234 
Corresponding status-go PR: https://github.com/status-im/status-go/pull/3066

### What does the PR do

* Support ENS usernames on different networks.
* Only current network ENS usernames are shown in settings.
* The new table is maintained fully in status-go, not from app (as it was before)
* Preferred username is temporarily dumb-fixed during network chain: https://github.com/status-im/status-desktop/issues/9035

### Affected areas

ENS usernames settings

### Screenshot of functionality (including design for comparison)

https://user-images.githubusercontent.com/25482501/211618608-3a0654fa-72c8-4fa9-a146-3fd88ad9640a.mov

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 93429ce3 | [#1](https://ci.status.im/job/status-desktop/job/branches/job/imports/job/PR-9036/1/) | 2023-01-10 17:23:59 | ~5 min | `imports` | [:page_facing_up:`log`](https://ci.status.im/job/status-desktop/job/branches/job/imports/job/PR-9036/1//consoleText) |

anastasiyaig commented:
@igor-sirotin can u please restart nim tests and gather the review please? The pr seems to be there for quite a while :(

alaibe commented on .gitignore:
```diff
@@ -1,4 +1,6 @@
 /Status
+/Status copy
+/Status backup
```
nit: that seems specific to u and not the project, maybe remove those 2 lines

alaibe commented on password-hash.py:
```diff
@@ -0,0 +1,12 @@
+hash = '0x' + hasher.hexdigest().upper()
+pyperclip.copy(hash)
+
+print(f'Hash: {hash} is copied to clipboard')
```
useful to add in scripts directory?

alaibe commented on src/app/modules/main/profile_section/ens_usernames/controller.nim:
```diff
@@ -63,44 +63,47 @@ proc init*(self: Controller) =
+proc setPubKey*(self: Controller, chainId: int, ensUsername: string, address: string, gas: string, gasPrice: string,
   maxPriorityFeePerGas: string, maxFeePerGas: string, password: string, eip1559Enabled: bool): string =
-  return self.ensService.setPubKey(ensUsername, address, gas, gasPrice, maxPriorityFeePerGas, maxFeePerGas, password, eip1559Enabled)
+  return self.ensService.setPubKey(chainId, ensUsername, address, gas, gasPrice, maxPriorityFeePerGas, maxFeePerGas, password, eip1559Enabled)
```
why do u need to pass the chain ID down to the service while it could be resolved there? *question valid for most of the function here

alaibe replied:
I think you're right, this is not a bad idea

alaibe commented on src/app_service/service/ens/service.nim:
```diff
@@ -140,16 +146,58 @@ QtObject:
+      for dto in self.getMyPendingEnsUsernames():
+        result.add(dto)
+
+  proc add*(self: Service, chainId: int, username: string): bool =
```
why do we need public api for add / remove? 
This should only be handle via register/set public and release?

alaibe replied:
`connectOwnedUsername` is it because u reboot your status folder?
`remove` is it to just not display a ens?

alaibe commented on src/app_service/service/ens/service.nim:
```diff
@@ -217,20 +267,17 @@ QtObject:
+  proc setPubKeyGasEstimate*(self: Service, chainId: int, ensUsername: string, address: string): int = 
     try:
-      let
-        chainId = self.networkService.getNetworkForEns().chainId
```
That looks more logical to not allow the service to not be called by any chain id

alaibe reviewed (commented):
I mostly don't understand why we must pass chain id down while it is static based on the fact that you are either in prod or test net?

richard-ramos commented on src/app/modules/main/profile_section/ens_usernames/view.nim:
```diff
@@ -39,6 +39,16 @@ QtObject:
+    read = chainId
+    notify = chainIdChanged
+
+  proc emitChaindIdChanged*(self: View) =
```
```suggestion
  proc emitChainIdChanged*(self: View) =
```

richard-ramos reviewed (approved):
:medal_sports:

jrainville reviewed (approved):
LGTM.

Don't forget to assign a QA to test it out

anastasiyaig commented:
even better, i will ask @elina2015 to test it today! :) Elina, please:)

elina2015 reviewed (changes requested):
I installed the mac build on two machines and have the same result - after installing and signing up for a new account, after the page of a new password and its confirmation, the app is closed. 
@igor-sirotin  could you kindly check?

igor-sirotin commented:
@elina2015, please take a look, I fixed that 🙂

elina2015 commented:
@igor-sirotin thank you, now the registration of a new account works well, but the app crashed when I'm going to buy an ENS name.
Steps:
1. I activated the test mode and added test tokens
2. I enter a valid and available ENS name for purchasing and after clicking on the button 'Next' the app crashed and closed.

elina2015 reviewed (changes requested):
@igor-sirotin please check the crash after trying to buy a n ENS name

igor-sirotin commented:
@elina2015 

First of all, I'm sorry, I should've tested it again myself before giving back to you.

I've checked it now, the bug is coming not from my changes 🙁 
I opened a separate issue on this, can be tracked here: https://github.com/status-im/status-desktop/issues/9264
I will start fixing it now, probably there's no way to test this PR before fixing the bug.

anastasiyaig commented:
@igor-sirotin sadly @elina2015 requested changes and she has to approve now to let this PR be merged
@elina2015 please approve

igor-sirotin commented:
@anastasiyaig what happened? why is this?
There's still a bug, I can confirm it and am working on it to fix.

anastasiyaig commented:
ah, i overlooked then! I might misread the comments thread , sorry about that @igor-sirotin

igor-sirotin commented:
@elina2015, I've fixed that bug here https://github.com/status-im/status-desktop/pull/9330 and rebased this PR on required changes.
You can test it now, but please don't merge, I'will rebase it to remove that commit (as it's part of another PR).

igor-sirotin commented:
The underlying patch merged, I removed the commit from this branch. 
Merging is ok now, waiting for testing.
